src/psyclone/tests/dynamo0p3_transformations_test.py

```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+
+    # start after end
+    with pytest.raises(TransformationError) as excinfo:
+        schedule, _ = mtrans.apply(schedule.children[1],
+                                   schedule.children[3])
+    assert "dependencies forbid" in str(excinfo.value)
+
+
+def test_async_hex_move_error_2():
+    '''Test that an asynchronous halo exchange start can not be moved
+    before a kernel that modifies the field and its end cannot be
+    moved after a kernel that reads the field
+
+    '''
+    _, invoke_info = parse(os.path.join(
+        os.path.dirname(os.path.abspath(__file__)),
+        "test_files", "dynamo0p3",
+        "4.3_multikernel_invokes.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    schedule = psy.invokes.invoke_list[0].schedule
+
+    f1_hex = schedule.children[5]
+    ahex_trans = Dynamo0p3AsyncHaloExchangeTrans()
+    schedule, _ = ahex_trans.apply(f1_hex)
+
+    mtrans = MoveTrans()
+
+    # start before prev modifier
+    with pytest.raises(TransformationError) as excinfo:
+        schedule, _ = mtrans.apply(schedule.children[5],
+                                   schedule.children[4])
+    assert "dependencies forbid" in str(excinfo.value)
+
+    # end after following reader
+    with pytest.raises(TransformationError) as excinfo:
+        schedule, _ = mtrans.apply(schedule.children[6],
+                                   schedule.children[7],
+                                   position="after")
+    assert "dependencies forbid" in str(excinfo.value)
+
+
+def test_rc_remove_async_halo_exchange(monkeypatch, tmpdir, f90, f90flags):
+    '''Test that a halo exchange is removed if redundant computation means
+    that it is no longer required. Halo exchanges are not required in
+    this example when we compute annexed dofs. Therefore we ensure we
+    compute over owned dofs (via monkeypatch) to perform the test.
+
+    '''
+    monkeypatch.setattr(_API_CONFIG, "_compute_annexed_dofs", False)
+    _, info = parse(os.path.join(
+        BASE_PATH, "14.7_halo_annexed.f90"),
+                    api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(info)
+    schedule = psy.invokes.invoke_list[0].schedule
+
+    ahex_trans = Dynamo0p3AsyncHaloExchangeTrans()
+
+    f2_hex = schedule.children[3]
+    schedule, _ = ahex_trans.apply(f2_hex)
+    f1_hex = schedule.children[2]
+    schedule, _ = ahex_trans.apply(f1_hex)
+
+    result = str(psy.gen)
+    assert "CALL f1_proxy%halo_exchange_start(depth=1)" in result
+    assert "CALL f1_proxy%halo_exchange_finish(depth=1)" in result
+    assert "CALL f2_proxy%halo_exchange_start(depth=1)" in result
+    assert "CALL f2_proxy%halo_exchange_finish(depth=1)" in result
+    assert "IF (m1_proxy%is_dirty(depth=1)) THEN" in result
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+    #
+
+    invoke = psy.invokes.invoke_list[0]
```

Comment:
You've already got a reference to `schedule` above.

Reply:
Removed.